analyses/cms-open-data-ttbar/validate_histograms.py

```diff
@@ -0,0 +1,74 @@
+# Compare the content of histograms produced by ttbar_analysis_pipeline with a reference file.
+# A reference file for N_FILES_MAX_PER_SAMPLE=1 is available in directory `reference/`.
+
+import argparse
+from collections import defaultdict
+import json
+import numpy as np
+import sys
+import uproot
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--histos", help="ROOT file containing the output histograms. Defaults to './histograms.root'.", default="histograms.root")
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument("--reference", help="JSON reference against which histogram contents should be compared")
+    group.add_argument("--dump-json", help="Print JSON representation of histogram contents to screen", action='store_true')
+    return parser.parse_args()
+
+# convert uproot file containing only TH1Ds to a corresponding JSON-compatible dict with structure:
+# { "histo1": { "edges": [...], "contents": [...] }, "histo2": { ... }, ... }
+# Only the highest namecycle for every histogram is considered, and cycles are stripped from the histogram names.
+def as_dict(f: uproot.ReadOnlyDirectory) -> dict[str, dict]:
+    histos = defaultdict(dict)
+    cycles: dict(str, int) = {}
+    for k, v in f.items():
+        assert isinstance(v, uproot.behaviors.TH1.Histogram)
+        # this assumes that the rightmost ";" (if any) comes before a namecycle
+        name, *cycle = k.rsplit(";", 1)
+        cycle = int(cycle[0]) if len(cycle) > 0 else -1
+        if name in histos and cycles["cycle"] > cycle:
+            continue # found a lower cycle for a histogram we already have
+
+        cycles[name] = cycle
+        histos[name]["edges"] = v.axis().edges().tolist()
+        histos[name]["contents"] = v.counts(flow=True).tolist()
+    return histos
+
+def validate(histos: dict, reference: dict) -> dict[str, list[str]]:
```

Comment:
This requires Python 3.9+, so I would suggest to add an `from __future__ import annotations` for backwards compatibility at the beginning of the file. While [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) puts us already in the 3.9+ age, I imagine not everyone is up-to-date with that yet.

Reply:
I agree for this repo, but am hoping that this might be useful more broadly and the cost of the extra line is negligible for this case.